### PR TITLE
add cleos convert pack_transaction and unpack_transaction

### PR DESCRIPTION
**Pull request by eosbase**
I needed a way to pack and unpack transactions from shell scripts and found this easiest to do.

Would something like this be accepted ? Or are any changes required ?

Thanks, Leon

--

for example pack:
```
# cleos convert pack_transaction '{
  "expiration": "2018-08-02T20:24:36",
  "ref_block_num": 14207,
  "ref_block_prefix": 1438248607,
  "max_net_usage_words": 0,
  "max_cpu_usage_ms": 0,
  "delay_sec": 0,
  "context_free_actions": [],
  "actions": [{
      "account": "eosio",
      "name": "newaccount",
      "authorization": [{
          "actor": "eosio",
          "permission": "active"
        }
      ],
      "data": "0000000000ea305500a6823403ea30550100000001000240cc0bf90a5656c8bb81f0eb86f49f89613c5cd988c018715d4646c6bd0ad3d8010000000100000001000240cc0bf90a5656c8bb81f0eb86f49f89613c5cd988c018715d4646c6bd0ad3d801000000"
    }
  ],
  "transaction_extensions": []
}'

{
  "signatures": [],
  "compression": "none",
  "packed_context_free_data": "",
  "packed_trx": "8468635b7f379feeb95500000000010000000000ea305500409e9a2264b89a010000000000ea305500000000a8ed3232660000000000ea305500a6823403ea30550100000001000240cc0bf90a5656c8bb81f0eb86f49f89613c5cd988c018715d4646c6bd0ad3d8010000000100000001000240cc0bf90a5656c8bb81f0eb86f49f89613c5cd988c018715d4646c6bd0ad3d80100000000"
}
```

and unpack:
```
# cleos convert unpack_transaction '{
  "signatures": [
    "SIG_K1_KmRbWahefwxs6uyCGNR6wNRjw7cntEeFQhNCbyg8S92Kbp7zdSSVGTD2QS7pNVWgcU126zpxaBp9CwUxFpRwSnfkjd46bS"
  ],
  "compression": "none",
  "packed_context_free_data": "",
  "packed_trx": "8468635b7f379feeb95500000000010000000000ea305500409e9a2264b89a010000000000ea305500000000a8ed3232660000000000ea305500a6823403ea30550100000001000240cc0bf90a5656c8bb81f0eb86f49f89613c5cd988c018715d4646c6bd0ad3d8010000000100000001000240cc0bf90a5656c8bb81f0eb86f49f89613c5cd988c018715d4646c6bd0ad3d80100000000"
}'

{
  "expiration": "2018-08-02T20:24:36",
  "ref_block_num": 14207,
  "ref_block_prefix": 1438248607,
  "max_net_usage_words": 0,
  "max_cpu_usage_ms": 0,
  "delay_sec": 0,
  "context_free_actions": [],
  "actions": [{
      "account": "eosio",
      "name": "newaccount",
      "authorization": [{
          "actor": "eosio",
          "permission": "active"
        }
      ],
      "data": "0000000000ea305500a6823403ea30550100000001000240cc0bf90a5656c8bb81f0eb86f49f89613c5cd988c018715d4646c6bd0ad3d8010000000100000001000240cc0bf90a5656c8bb81f0eb86f49f89613c5cd988c018715d4646c6bd0ad3d801000000"
    }
  ],
  "transaction_extensions": []
}
```

### CODE REVIEW

**Comment by eosbase** [comment]
thanks, I will do that!

File: programs/cleos/main.cpp [hunk 1]
```diff
@@ -1725,6 +1752,86 @@ int main( int argc, char** argv ) {
    // create account
    auto createAccount = create_account_subcommand( create, true /*simple*/ );
 
+   // convert subcommand
+   auto convert = app.add_subcommand("convert", localized("Pack and unpack transactions"), false); // TODO also add converting action args based on abi from here ?
```
Comment (heifner) [inline programs/cleos/main.cpp]: To address your TODO, I do think having the ability to specify an abi file to use would be nice. In that case it should NOT connect to nodeos or need to.

File: programs/cleos/main.cpp [hunk 2]
```diff
@@ -1725,6 +1752,86 @@ int main( int argc, char** argv ) {
...
+      } EOS_RETHROW_EXCEPTIONS(transaction_type_exception, "Fail to parse packed transaction JSON '${data}'", ("data", packed_transaction_json))
+      string trx_hex;
+      try {
+         trx_hex = trx_var["packed_trx"].as_string();
+      } EOS_RETHROW_EXCEPTIONS(transaction_type_exception, "Missing packed_trx field in provided JSON '${data}'", ("data", packed_transaction_json))
+
+      std::vector<char> trx_blob(trx_hex.size()/2);
+      fc::from_hex(trx_hex, trx_blob.data(), trx_blob.size());
+      transaction unpacked_trx = fc::raw::unpack<transaction>(trx_blob);
+      // TODO would it be nice to put the unpacked_trx actions into the trx actions in a separate unpacked_data field  or somesuch ?
```
Comment (heifner) [inline programs/cleos/main.cpp]: Remove TODO, that would require api changes.

File: programs/cleos/main.cpp [hunk 3]
```diff
@@ -1725,6 +1752,86 @@ int main( int argc, char** argv ) {
...
+      if (unpack_action_data_flag) {
+         for ( const auto& a : unpacked_trx.actions ) {
+            fc::variant unpacked_action_data = bin_to_variant(a.account, a.name, a.data);
+            std::cout << fc::json::to_pretty_string(unpacked_action_data) << std::endl;
+         }
+      } else {
+	 // TODO should the "sigantures" and "context_free_data" fields be copied from trx_var into the action field(s) ?
```
Comment (heifner) [inline programs/cleos/main.cpp]: yes, signatures and context_free_data should be added to the signed transaction if available.

File: programs/cleos/main.cpp [hunk 4]
```diff
@@ -1725,6 +1752,86 @@ int main( int argc, char** argv ) {
...
+      signed_transaction trx = trx_var.as<signed_transaction>();
+      std::cout << fc::json::to_pretty_string(fc::variant(packed_transaction(trx, packed_transaction::none))) << std::endl;
+   });
+
+   // unpack transaction
+   string packed_transaction_json;
+   bool unpack_action_data_flag = false;
+   auto unpack_transaction = convert->add_subcommand("unpack_transaction", localized("From packed to plain signed json form"));
+   unpack_transaction->add_option("transaction", packed_transaction_json, localized("The packed transaction json (string containing packed_trx and optionally compression fields)"))->required();
+   unpack_transaction->add_flag("--unpack-action-data", unpack_action_data_flag, localized("Upack all action datas within transaction and only return those, needs interaction with nodeos"));
```
Comment (heifner) [inline programs/cleos/main.cpp]: Not sure why anyone would use `--unpack-action-data`. What is the use case? I think it is fine, just not sure anyone would use it.

**Comment by heifner** [comment]
Replaced by #5225

**Pull request closed without merging**